src/expr/src/relation/join_input_mapper.rs

```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
         }
         None
     }
+
+    /// Try to find a consequence `c` of the given expression `e` for the given input.
+    ///
+    /// If we return `Some(c)`, that means
+    ///   1. `c` uses only columns from the given input;
+    ///   2. if `c` doesn't hold on a row of the input then `e` also wouldn't hold;
+    ///   3. if `c` holds on a row of the input then `e` might or might not hold.
+    /// 1. and 2. means that if we have a join with predicate `e` then we can use `c` for
+    /// pre-filtering a join input before the join. However, 3. means that `e` shouldn't be deleted
+    /// from the join predicates, i.e., we can't do a "traditional" predicate pushdown.
+    ///
+    /// Note that "`c` is a consequence of `e`" is the same thing as 2., see
+    /// <https://en.wikipedia.org/wiki/Contraposition>
+    ///
+    /// Example: For
+    /// `(t1.f2 = 3 AND t2.f2 = 4) OR (t1.f2 = 5 AND t2.f2 = 6)`
+    /// we find
+    /// `t1.f2 = 3 OR t1.f2 = 5` for t1, and
+    /// `t2.f2 = 4 OR t2.f2 = 6` for t2.
+    ///
+    /// Further examples are TPC-H Q07, Q19, and chbench Q07, Q19.
```

Comment:
Please explain the parameters. Thanks for adding such a detailed description of what the function does!

Reply:
Done